File: packages/tx/src/tx.ts
// @bsv-universal/tx — real BSV transaction model (REQ-SEC-007): canonical serialization + txid.
// Integers are bigint/number with explicit range checks (no float; CWE-190 guarded). All structure
// is bounded (MAX_IO) so a hostile tx cannot exhaust resources.

import { createHash } from 'node:crypto';
import { concatBytes, u32be } from '@bsv-universal/protocol-types';

export const MAX_IO = 10_000; // bound inputs/outputs per tx (CWE-770)

export interface Outpoint {
  readonly txid: Uint8Array; // 32 bytes, internal byte order
  readonly vout: number; // uint32
}
export interface TxInput {
  readonly outpoint: Outpoint;
  readonly unlockingScript: Uint8Array;
  readonly sequence: number; // uint32 (REQ-TIME: relative-locktime carrier; no CSV opcode)
}
export interface TxOutput {
  readonly satoshis: bigint;
  readonly lockingScript: Uint8Array;
}
export interface Tx {
  readonly version: number; // uint32
  readonly inputs: readonly TxInput[];
  readonly outputs: readonly TxOutput[];
  readonly lockTime: number; // uint32 (REQ-TIME: absolute-locktime carrier; no CLTV opcode)
}

function u32le(n: number): Uint8Array {
  if (!Number.isInteger(n) || n < 0 || n > 0xffffffff) throw new Error(`u32 out of range: ${n}`);
  return new Uint8Array([n & 0xff, (n >>> 8) & 0xff, (n >>> 16) & 0xff, (n >>> 24) & 0xff]);
}
function u64le(n: bigint): Uint8Array {
  if (typeof n !== 'bigint' || n < 0n || n > 0xffffffffffffffffn) throw new Error(`u64 out of range: ${n}`);
  const out = new Uint8Array(8);
  let v = n;
  for (let i = 0; i < 8; i++) {
    out[i] = Number(v & 0xffn);
    v >>= 8n;
  }
  return out;
}

/** Bitcoin CompactSize varint. */
export function varint(n: number): Uint8Array {
  if (!Number.isInteger(n) || n < 0) throw new Error(`varint out of range: ${n}`);
  if (n < 0xfd) return new Uint8Array([n]);
  if (n <= 0xffff) return new Uint8Array([0xfd, n & 0xff, (n >>> 8) & 0xff]);
  if (n <= 0xffffffff) return concatBytes(new Uint8Array([0xfe]), u32le(n));
  return concatBytes(new Uint8Array([0xff]), u64le(BigInt(n)));
}

function varBytes(b: Uint8Array): Uint8Array {
  return concatBytes(varint(b.length), b);
}

export const sha256 = (b: Uint8Array): Uint8Array => new Uint8Array(createHash('sha256').update(b).digest());
export const hash256 = (b: Uint8Array): Uint8Array => sha256(sha256(b));

function checkBounds(tx: Tx): void {
  if (!Number.isInteger(tx.version) || tx.version < 0 || tx.version > 0xffffffff) throw new Error('bad version');
  if (!Number.isInteger(tx.lockTime) || tx.lockTime < 0 || tx.lockTime > 0xffffffff) throw new Error('bad lockTime');
  if (tx.inputs.length > MAX_IO || tx.outputs.length > MAX_IO) throw new Error('too many inputs/outputs');
}

/** Canonical transaction serialization. */
export function serializeTx(tx: Tx): Uint8Array {
  checkBounds(tx);
  const parts: Uint8Array[] = [u32le(tx.version), varint(tx.inputs.length)];
  for (const i of tx.inputs) {
    if (i.outpoint.txid.length !== 32) throw new Error('outpoint txid must be 32 bytes');
    parts.push(i.outpoint.txid, u32le(i.outpoint.vout), varBytes(i.unlockingScript), u32le(i.sequence));
  }
  parts.push(varint(tx.outputs.length));
  for (const o of tx.outputs) parts.push(u64le(o.satoshis), varBytes(o.lockingScript));
  parts.push(u32le(tx.lockTime));
  return concatBytes(...parts);
}

/** Transaction id = double-SHA256 of the canonical serialization (internal byte order). */
export function txid(tx: Tx): Uint8Array {
  return hash256(serializeTx(tx));
}

export { u32le, u64le, varBytes, u32be };

File: go/txmodel/tx.go
// Package txmodel is the independent Go implementation of the BSV transaction model (REQ-TEST-003
// extended to the value layer). It MUST be byte-identical to @bsv-universal/tx: same serialization,
// same txid (double-SHA256), same BIP143/forkid sighash, same value conservation, same covenant
// binding. The differential proves it.
package txmodel

import (
	"crypto/sha256"
	"math/big"
)

const sighashAllForkid = 0x41

type Outpoint struct {
	Txid []byte // 32 bytes
	Vout uint32
}
type TxInput struct {
	Outpoint        Outpoint
	UnlockingScript []byte
	Sequence        uint32
}
type TxOutput struct {
	Satoshis      *big.Int
	LockingScript []byte
}
type Tx struct {
	Version uint32
	Inputs  []TxInput
	Outputs []TxOutput
	LockTime uint32
}

func sha256d(b []byte) []byte {
	a := sha256.Sum256(b)
	c := sha256.Sum256(a[:])
	return c[:]
}

func u32le(n uint32) []byte {
	return []byte{byte(n), byte(n >> 8), byte(n >> 16), byte(n >> 24)}
}
func u64le(n *big.Int) []byte {
	v := n.Uint64()
	out := make([]byte, 8)
	for i := 0; i < 8; i++ {
		out[i] = byte(v)
		v >>= 8
	}
	return out
}

// Varint is the Bitcoin CompactSize encoding.
func Varint(n int) []byte {
	switch {
	case n < 0xfd:
		return []byte{byte(n)}
	case n <= 0xffff:
		return []byte{0xfd, byte(n), byte(n >> 8)}
	case n <= 0xffffffff:
		return append([]byte{0xfe}, u32le(uint32(n))...)
	default:
		return append([]byte{0xff}, u64le(big.NewInt(int64(n)))...)
	}
}

func varBytes(b []byte) []byte {
	return append(Varint(len(b)), b...)
}

func cat(parts ...[]byte) []byte {
	var out []byte
	for _, p := range parts {
		out = append(out, p...)
	}
	return out
}

// Serialize is the canonical transaction serialization.
func Serialize(tx Tx) []byte {
	out := cat(u32le(tx.Version), Varint(len(tx.Inputs)))
	for _, in := range tx.Inputs {
		out = cat(out, in.Outpoint.Txid, u32le(in.Outpoint.Vout), varBytes(in.UnlockingScript), u32le(in.Sequence))
	}
	out = cat(out, Varint(len(tx.Outputs)))
	for _, o := range tx.Outputs {
		out = cat(out, u64le(o.Satoshis), varBytes(o.LockingScript))
	}
	return cat(out, u32le(tx.LockTime))
}

// Txid = double-SHA256(serialize), internal byte order.
func Txid(tx Tx) []byte { return sha256d(Serialize(tx)) }

func hashPrevouts(tx Tx) []byte {
	var b []byte
	for _, in := range tx.Inputs {
		b = cat(b, in.Outpoint.Txid, u32le(in.Outpoint.Vout))
	}
	return sha256d(b)
}
func hashSequence(tx Tx) []byte {
	var b []byte
	for _, in := range tx.Inputs {
		b = cat(b, u32le(in.Sequence))
	}
	return sha256d(b)
}
func hashOutputs(tx Tx) []byte {
	var b []byte
	for _, o := range tx.Outputs {
		b = cat(b, u64le(o.Satoshis), varBytes(o.LockingScript))
	}
	return sha256d(b)
}

// SighashPreimage is the BIP143 preimage for input `index` (SIGHASH_ALL|FORKID).
func SighashPreimage(tx Tx, index int, prevScript []byte, amount *big.Int) []byte {
	in := tx.Inputs[index]
	return cat(
		u32le(tx.Version),
		hashPrevouts(tx),
		hashSequence(tx),
		in.Outpoint.Txid,
		u32le(in.Outpoint.Vout),
		varBytes(prevScript),
		u64le(amount),
		u32le(in.Sequence),
		hashOutputs(tx),
		u32le(tx.LockTime),
		u32le(sighashAllForkid),
	)
}

// Sighash = double-SHA256(preimage).
func Sighash(tx Tx, index int, prevScript []byte, amount *big.Int) []byte {
	return sha256d(SighashPreimage(tx, index, prevScript, amount))
}

// VerifyTxValue conserves value against real previous amounts + fee. Returns (ok, reason).
func VerifyTxValue(tx Tx, prevAmounts []*big.Int, fee *big.Int) (bool, string) {
	if len(prevAmounts) != len(tx.Inputs) {
		return false, "prevAmounts must match inputs"
	}
	if fee.Sign() < 0 {
		return false, "fee must be non-negative"
	}
	totalIn := big.NewInt(0)
	for _, v := range prevAmounts {
		if v.Sign() < 0 {
			return false, "prev amount must be non-negative"
		}
		totalIn.Add(totalIn, v)
	}
	totalOut := big.NewInt(0)
	for _, o := range tx.Outputs {
		if o.Satoshis.Sign() < 0 {
			return false, "output satoshis must be non-negative"
		}
		totalOut.Add(totalOut, o.Satoshis)
	}
	if totalIn.Cmp(new(big.Int).Add(totalOut, fee)) != 0 {
		return false, "value not conserved"
	}
	return true, "ok"
}
